src/rdf4cpp/rdf/datatypes/xsd/Base64Binary.cpp

```diff
@@ -0,0 +1,184 @@
+
+        while (ch == ' ') {
+            assert(pos < base64encoded.size());
+            ch = base64encoded[pos++];
+        }
+
+        if (pos - old_pos > 2) {
+            throw std::runtime_error{"xsd:base64Binary parsing error: too many consecutive spaces"};
+        }
+
+        return ch;
+    };
+
+    std::vector<std::byte> buf;
+    buf.reserve(base64encoded.size() / 4 * 3);
+
+    for (size_t quad = 0; quad < len / 4 - 1; ++quad) {
+        auto const c1 = next_char();
+        auto const c2 = next_char();
+        auto const c3 = next_char();
+        auto const c4 = next_char();
+
+        auto const decoded = encode_decode_detail::base64_decode(c1, c2, c3, c4);
+        std::copy(decoded.begin(), decoded.end(), std::back_inserter(buf));
+    }
+
+    {
+        auto const c1 = next_char();
+        auto const c2 = next_char();
+        auto const c3 = next_char();
+        auto const c4 = next_char();
+
+        auto const c3_pad = c3 == '=';
+        auto const c4_pad = c4 == '=';
+
+        if (c3_pad && !c4_pad) {
+            throw std::runtime_error{"xsd:binary64 parsing error: Invalid padding"};
+        }
+
+        size_t const n = c3_pad + c4_pad;
+        auto const decoded = encode_decode_detail::base64_decode(c1, c2, c3_pad ? 'A' : c3, c4_pad ? 'A' : c4);
+        std::copy_n(decoded.begin(), 3 - n, std::back_inserter(buf));
+    }
+
+    return Base64BinaryRepr{buf};
+}
+
+std::string Base64BinaryRepr::to_encoded() const noexcept {
+    if (this->empty()) {
+        return "";
+    }
+
+    std::string buf;
+    for (size_t triple = 0; triple < this->size() / 3; ++triple) {
+        auto const b1 = (*this)[triple * 3];
+        auto const b2 = (*this)[triple * 3 + 1];
+        auto const b3 = (*this)[triple * 3 + 2];
+
+        auto const encoded = encode_decode_detail::base64_encode(b1, b2, b3);
+        std::copy(encoded.begin(), encoded.end(), std::back_inserter(buf));
+    }
+
+    if (auto const rest = this->size() % 3; rest != 0) {
+        auto const triple = this->size() / 3;
+
+        auto const b1 = (*this)[triple * 3];
+        auto const b2 = triple * 3 + 1 < this->size() ? (*this)[triple * 3 + 1] : std::byte{0};
+        auto const b3 = triple * 3 + 2 < this->size() ? (*this)[triple * 3 + 2] : std::byte{0};
+
+        auto const encoded = encode_decode_detail::base64_encode(b1, b2, b3);
+        std::copy_n(encoded.begin(), 4 - (3 - rest), std::back_inserter(buf));
+        std::fill_n(std::back_inserter(buf), 3 - rest, '=');
+    }
+
+    return buf;
+}
+
+std::byte Base64BinaryRepr::hextet(size_t const n) const noexcept {
+    auto const triple = n / 4 * 3;
+    auto const off = (3 - (n % 4)) * 6;
+
+    uint32_t const bytes = static_cast<uint32_t>((*this)[triple]) << 16
+                           | (triple + 1 < this->size() ? static_cast<uint32_t>((*this)[triple + 1]) << 8 : 0)
+                           | (triple + 2 < this->size() ? static_cast<uint32_t>((*this)[triple + 2]) : 0);
```

Comment:
```suggestion
    uint32_t const bytes = (triple < this->size() ? static_cast<uint32_t>((*this)[triple]) << 16 : 0)
                           | (triple + 1 < this->size() ? static_cast<uint32_t>((*this)[triple + 1]) << 8 : 0)
                           | (triple + 2 < this->size() ? static_cast<uint32_t>((*this)[triple + 2]) : 0);
```

Or should we do bound-checking here and throw an error?

Reply:
The reason I did it this way is that `n_hextets()` returns the number of hextets in the encoded string
meaning if there are padding chars '=' they also count towards it.

Example: `Hello World` <-> `SGVsbG8gV29ybGQ=`

In this value you can currently extract the last/padding hextet ('='). You get a value of zero for it.
But the type itself does not store the corresponding bytes that contain the padding hextets, as these bytes are by definition just zero bytes that were appended to the data to make it encodable.

Meaning: `n_hextets() == len(encoded value) == 16`  but `n_bytes() == len(decoded value) == 11`

Therefore to extract the padding hextet, the padding byte that does not exist in the decoded value needs to be
artificially inserted to not have a buffer overread.

Reason for why I chose this behaviour:
- the encoded padding hextets are required and have a purpose in the encoded value
- the padding bytes in the decoded value however do not have any meaning (apart from being padding) and were not part of the original data

Although I would agree that it is not the most efficient way to implement that (because of doing bounds checking twice). It was the least ugly way I could come up with.

Reply:
Thx for the explanation. Makes sense.

The one thing that still bothers me is what happens if triple >= this->size(). This is a index out of bound. Behavior is principally fine but needs to be documented.

Reply:
added warnings to index functions